### PR TITLE
[ie/bandcamp] Extract extended user page

### PR DESCRIPTION
**Pull request by quad**
### Description of your *pull request* and other information

The user page is lazy-loaded from the JSON embedded in the data-client-items attribute. Let's use it, if it exists.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)

</details>

### CODE REVIEW

**Comment by quad** [comment]
@bashonly AFAIK this PR is ready to merge. Is there anything else I need to do?

File: yt_dlp/extractor/bandcamp.py [hunk 1]
```diff
@@ -473,12 +473,20 @@ class BandcampUserIE(InfoExtractor):
         },
     }]
 
+    def _extract_data_attr(self, webpage, video_id, attr, fatal=True):
+        return self._parse_json(self._html_search_regex(
+            fr'data-{attr}=(["\'])(\[.+?\])\1', webpage,
+            attr + ' data', group=2), video_id, fatal=fatal)
```
Comment (bashonly) [inline yt_dlp/extractor/bandcamp.py:479]: I would instead use a generator method to yield all the url paths
```suggestion
    def _yield_items(self, webpage):
        yield from (
            re.findall(r'<li data-item-id=["\'][^>]+>\s*<a href=["\'](?![^"\'/]*?/merch)([^"\']+)', webpage)
            or re.finditer(r'<div[^>]+trackTitle["\'][^"\']+["\']([^"\']+)', webpage))

        yield from traverse_obj(webpage, (
            {functools.partial(get_element_html_by_id, 'music-grid')}, {extract_attributes},
            'data-client-items', {json.loads}, ..., 'page_url', {str}))
```

File: yt_dlp/extractor/bandcamp.py [hunk 2]
```diff
@@ -473,12 +473,20 @@ class BandcampUserIE(InfoExtractor):
...
+        matches = []
+        if discography_data := (re.findall(r'<li data-item-id=["\'][^>]+>\s*<a href=["\'](?![^"\'/]*?/merch)([^"\']+)', webpage)
+                                or re.findall(r'<div[^>]+trackTitle["\'][^"\']+["\']([^"\']+)', webpage)):
+            matches += (urljoin(url, path) for path in discography_data)
+        if music_grid := self._extract_data_attr(webpage, uploader, 'client-items', fatal=False):
+            matches += (urljoin(url, album.get('page_url')) for album in music_grid)
 
-        return self.playlist_from_matches(
-            discography_data, uploader, f'Discography of {uploader}', getter=lambda x: urljoin(url, x))
+        return self.playlist_from_matches(matches, uploader, f'Discography of {uploader}')
```
Comment (bashonly) [inline yt_dlp/extractor/bandcamp.py:492]: and then do this
```suggestion
        return self.playlist_from_matches(
            self._yield_items(webpage), uploader, f'Discography of {uploader}',
            getter=functools.partial(urljoin, url))
```

File: yt_dlp/extractor/bandcamp.py [hunk 3]
```diff
@@ -473,12 +473,20 @@ class BandcampUserIE(InfoExtractor):
...
         webpage = self._download_webpage(url, uploader)
 
-        discography_data = (re.findall(r'<li data-item-id=["\'][^>]+>\s*<a href=["\'](?![^"\'/]*?/merch)([^"\']+)', webpage)
-                            or re.findall(r'<div[^>]+trackTitle["\'][^"\']+["\']([^"\']+)', webpage))
+        matches = []
+        if discography_data := (re.findall(r'<li data-item-id=["\'][^>]+>\s*<a href=["\'](?![^"\'/]*?/merch)([^"\']+)', webpage)
+                                or re.findall(r'<div[^>]+trackTitle["\'][^"\']+["\']([^"\']+)', webpage)):
+            matches += (urljoin(url, path) for path in discography_data)
+        if music_grid := self._extract_data_attr(webpage, uploader, 'client-items', fatal=False):
+            matches += (urljoin(url, album.get('page_url')) for album in music_grid)
```
Comment (bashonly) [inline yt_dlp/extractor/bandcamp.py:491]:
```suggestion
```

**Review by bashonly — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing.

I think we can take a simpler approach, see review comments below.

In addition to my suggestions, you would need to import `functools` and `json`, import `extract_attributes` and `get_element_html_by_id` from `..utils`, and import `traverse_obj` from `..utils.traversal`.

Please also update the tests if their counts have changed, and add a test if the existing tests don't provide coverage for the extended extraction

**Comment by formtapez** [comment]
Any news on this?

**Comment by formtapez** [comment]
Actual example: `https://traumschallplatten.bandcamp.com`

**Review by bashonly — APPROVED** [review]

**Pull request merged**